File: TPs/TP1/E1.py
"""
Desarrollar una función que reciba tres números enteros positivos y devuelva
el mayor de los tres, sólo si éste es único (es decir el mayor estricto).
Devolver -1 en caso de no haber ninguno. No utilizar operadores lógicos
(and, or, not). Desarrollar también un programa para ingresar los tres
valores, invocar a la función y mostrar el máximo hallado, o un mensaje
informativo si éste no existe.
"""

from functools import reduce
from typing import List


def numeros_input() -> List[int]:
    """Pide y guarda 3 números positivos en una lista.

    Pre: No recibe nada.

    Post: Devuelve una lista que contiene exactamente tres números enteros
          positivos.

    """
    numeros = []
    while len(numeros) < 3:
        try:
            numero = input(f"Ingrese el número {len(numeros)+1}: ")
            if not numero.isdigit():
                raise ValueError()
            if numero > 0:
                numeros.append(numero)
            else:
                raise ValueError()
        except ValueError:
            print("Valor inválido, reintentar.")
            return []
    return numeros


def mayor(numeros: List[int]) -> int:
    """Busca el mayor estricto de los 3 y devuelve el número mayor o -1 si no es
    el mayor estricto.

    Pre: Recibe una lista de enteros con los 3 números mayores a 0.

    Post: Devuelve el mayor estricto si existe, si no devuelve -1.

    """
    n_mayor = reduce(lambda x, y: x if x > y else y, numeros)
    rep = numeros.count(n_mayor) - 1
    return n_mayor if rep == 0 else -1


def imprimir_resultados(resultado: int) -> None:
    """Imprime los resultados obtenidos de los procesos anteriores.

    Pre: Recibe el resultado de mayor(), un número entero.

    Post: Si resultado es distinto a -1 imprime el mayor estricto.

    """
    if resultado != -1:
        print(f"El número mayor ingresado es: {resultado}.")
    else:
        print("No hubo ningún número mayor estricto.")
    return None


def main() -> None:
    """Función principal del programa."""
    numeros = numeros_input()
    resultado = mayor(numeros)
    imprimir_resultados(resultado)
    return None


if __name__ == "__main__":
    main()

# End-of-file (EOF)

File: TPs/TP1/E1.rs
/*
Desarrollar una función que reciba tres números enteros positivos y devuelva el
mayor de los tres, sólo si éste es único (es decir el mayor estricto). Devolver -1 en
caso de no haber ninguno. No utilizar operadores lógicos (and, or, not). Desarrollar
también un programa para ingresar los tres valores, invocar a la función y mostrar
el máximo hallado, o un mensaje informativo si éste no existe.
*/
use std::io;

fn input_number(message: &str) -> Vec<u8> {
    let mut data = Vec::new();
    while data.len() < 3 {
        let n: u8;
        let mut str = String::new();
        println!("{} {}:", message, data.len() + 1);
        io::stdin()
            .read_line(&mut str)
            .expect("No se pudo leer el input.");
        match str.trim().parse::<u8>() {
            Ok(value) => {
                n = value;
                data.push(n);
            }
            Err(_) => {
                println!("Valor invalido, vuelva a intentarlo.");
            }
        }
    }
    data
}

fn greater(data: Vec<u8>) -> Result<u8, bool> {
    let n: u8;
    let count: u8;
    n = *data.iter().max().unwrap();
    count = data.iter().filter(|&&x| x == n).count() as u8;
    if count > 1 {
        Err(false)
    } else {
        Ok(n)
    }
}

fn main() {
    let data: Vec<u8> = input_number("Ingrese el numero");
    match greater(data) {
        Ok(n) => println!("El numero mayor estricto es: {}.", n),
        Err(_) => println!("No hubo numero mayor estricto."),
    }
}
